Update Amounts and Units for better sorting and filtering behavior

labkey-susanh opened the pull request:
#### Rationale
Our original implementation for the StoredAmount and Units columns added display columns over the user-provided data so the application grid displays used consistent units, but left the values as provided by users. Since we accept values in different, compatible units, this results in sorting and filtering on these columns not behaving as one would expect (e.g, samples with amounts of 2 mL and 2 L will be sorted next to each other, though displayed as 2 mL and 2000 mL in the grid).

This work update our handling of `StoredAmount` and `Units` so they always store values in the base units for sample types that have display units. This provides us with a valid basis for sorting and filtering. 

We also are updating the processing of data to always require that the user provides both StoredAmount and Units when either of these values is provided.

#### Related Pull Requests
- https://github.com/LabKey/platform/pull/6994 
- https://github.com/LabKey/limsModules/pull/1653
- https://github.com/LabKey/testAutomation/pull/2678
- https://github.com/LabKey/labkey-ui-components/pull/1850 
- https://github.com/LabKey/labkey-ui-premium/pull/813 

#### Changes
- Upgrade script to convert existing amounts and units to base units, and drop any units that don't have values
- Remove `Measurement` class in favor of `Quantity` and `Units` classes
- Update detailed audit log to handle provided values that are converted before being stored
- Update aliquot rollup logic to store values in base units
- Remove recompute of rollup values when display unit is changed
- Add `SampleTypeAmountDisplayColumn` and `SampleTyepUnitDisplayColumn` to display amount and units using chosen display units

labkey-matthewb committed:
quantity type prototype
(just working here, not for 24.11)

labkey-matthewb committed:
imports

labkey-matthewb committed:
Merge remote-tracking branch 'origin/develop' into fb_quantity

labkey-matthewb committed:
Merge remote-tracking branch 'origin/release24.11-SNAPSHOT' into 24.11_fb_quantity

labkey-matthewb committed:
merge patch

labkey-matthewb committed:
merge patch

labkey-matthewb committed:
java 17

labkey-matthewb committed:
merge forward

labkey-matthewb committed:
update expected result

labkey-matthewb committed:
build problem

labkey-matthewb committed:
I wasn't really planning on commiting this for a while.  But this makes conversion more consistent in a bunch of places.

ColumnRenderProperties.getConvertFn()

cnathe committed:
Merge branch 'develop' into fb_quantity

cnathe committed:
fixup after merge from develop

cnathe committed:
Merge branch 'develop' into fb_quantity

cnathe committed:
Add QUANTITY_COLUMN_SUFFIX_TESTING experimental feature flag

labkey-susanh committed:
Merge remote-tracking branch 'origin/develop' into fb_quantity

labkey-susanh committed:
Merge remote-tracking branch 'origin/develop' into fb_quantity

labkey-susanh committed:
Add "unit" unit to reflect current usage and some minor helper methods

cnathe committed:
Quantity.java unit tests

cnathe committed:
Set Unit enum base to "unit" and "ml"

cnathe committed:
fixup Quantity testFormat test case

cnathe committed:
Unit.java unit tests and set ml base unit value as 1 instead of liter

labkey-susanh committed:
Upgrade script to convert StoredAmount and Units and related aliquot fields to base units

labkey-susanh committed:
comment update

labkey-susanh committed:
Add null check in `isCompatible`

labkey-susanh committed:
compute aliquot total rollup in base units instead of display units

labkey-susanh committed:
Merge remote-tracking branch 'origin/develop' into fb_amountsAndUnits

# Conflicts:
#	api/src/org/labkey/api/data/ColumnRenderPropertiesImpl.java
#	api/src/org/labkey/api/data/DataColumn.java
#	api/src/org/labkey/api/data/DisplayColumn.java
#	api/src/org/labkey/api/ontology/KindOfQuantity.java
#	api/src/org/labkey/api/ontology/Quantity.java
#	api/src/org/labkey/api/ontology/Unit.java
#	experiment/resources/schemas/dbscripts/postgresql/exp-25.005-25.006.sql
#	experiment/resources/schemas/dbscripts/sqlserver/exp-25.005-25.006.sql

labkey-susanh committed:
Move to next schema version

labkey-susanh committed:
Add Nullable and NotNull annotations and getValidatedUnit method

labkey-susanh committed:
Use Quantity Unit instead of Measurement.Unit

labkey-susanh committed:
Remove unused (converted) convertToDisplayUnits method

labkey-susanh committed:
Add some @NotNull and @Nullable annotations

labkey-susanh committed:
Add utility methods for getting Unit objects for sample types

labkey-susanh committed:
AliquotUnit not Units

cnathe committed:
Unit enum update to use capital "L" for amount units (to match current app behavior) and to add precisionScale values

labkey-susanh committed:
Add hasUnit to Column

labkey-susanh committed:
make getStorageUnit public

labkey-susanh committed:
Another 'of' variant and add getter for KindOfQuantity

labkey-susanh committed:
Merge

labkey-susanh committed:
Precision

labkey-susanh committed:
Update conversion columns

labkey-susanh committed:
Merge remote-tracking branch 'origin/develop' into fb_amountsAndUnits

# Conflicts:
#	experiment/resources/schemas/dbscripts/postgresql/exp-25.006-25.007.sql
#	experiment/resources/schemas/dbscripts/sqlserver/exp-25.006-25.007.sql
#	experiment/src/org/labkey/experiment/ExperimentUpgradeCode.java

labkey-susanh committed:
Once more for the upgrade script

labkey-susanh committed:
Convert enum table to ontology Unit enum

labkey-susanh committed:
Change capitalization for liter-based units

labkey-susanh committed:
rawUnit may already be a Unit type object

labkey-susanh committed:
rawUnit may already be a Unit type object

labkey-susanh committed:
Handle null value

cnathe committed:
Remove AmountDataColumn.java and UnitsDataColumn.java

cnathe committed:
Add ExpMaterialTableImpl SampleTypeAmountDisplayColumn and SampleTypeUnitDisplayColumn as ExprColumns for amount/unit display

labkey-susanh committed:
Unit unit needs precision too

labkey-susanh committed:
Add override of coerceTypes to handle amount and unit columns when not doing data iteration. Also throw an exception if updating and not providing both the amount and the unit.

labkey-susanh committed:
Merge branch 'fb_amountsAndUnits' of https://github.com/LabKey/platform into fb_amountsAndUnits

labkey-susanh committed:
Merge remote-tracking branch 'origin/develop' into fb_amountsAndUnits

labkey-susanh committed:
display unit as default for amount not base unit

labkey-susanh committed:
null check

labkey-susanh committed:
Use label from enum

labkey-susanh committed:
Replace converted amount and unit values with raw values so audit difference is accurate.

labkey-susanh committed:
Merge remote-tracking branch 'origin/develop' into fb_amountsAndUnits

labkey-susanh committed:
Make pathway for provided values to be added to the audit log details and wire up values from non-DI operations

labkey-susanh committed:
Data iterator updates for capturing provided values

labkey-susanh committed:
Better handling for sample types without display units

labkey-susanh committed:
Remove use of deprecated method

cnathe committed:
MEASUREMENT_UNITS_TABLE to use KindOfQuantity getCommonUnits()

cnathe committed:
Merge branch 'develop' into fb_amountsAndUnits

labkey-susanh committed:
tidying

labkey-susanh committed:
Update logic for validating unit + amount columns

labkey-susanh committed:
Merge branch 'fb_amountsAndUnits' of https://github.com/LabKey/platform into fb_amountsAndUnits

labkey-susanh committed:
Merge remote-tracking branch 'origin/develop' into fb_amountsAndUnits

labkey-susanh committed:
Remove defaulting to sample type display unit and add handling for delta amounts when checking in samples

labkey-susanh committed:
Update test values

cnathe committed:
Merge branch 'develop' into fb_amountsAndUnits

labkey-susanh committed:
Merge remote-tracking branch 'origin/develop' into fb_amountsAndUnits

labkey-susanh committed:
Updates for providedData map to separate delta values to own map and fix up some key values in metadata map

labkey-susanh committed:
Remove unneeded code

labkey-susanh committed:
Remove Measurement.java

labkey-susanh committed:
Don't convert to base units if there is no display unit. Just store it as is

labkey-susanh committed:
Clean up confirmAmountAndUnitsColumn and coerceTypes override

cnathe committed:
Supported measurement units error to only list those common/supported units

labkey-susanh committed:
Update tests

labkey-susanh committed:
Update error message text

labkey-susanh committed:
Merge remote-tracking branch 'origin/develop' into fb_amountsAndUnits

cnathe committed:
Merge branch 'develop' into fb_amountsAndUnits

labkey-susanh committed:
Remove some fields from the audit logs details

labkey-susanh committed:
Don't update aliquot units if it's already in the base units

labkey-susanh committed:
More refinement for aliquots

labkey-susanh committed:
Don't use TableSelector since it uses the display value

cnathe committed:
Merge branch 'develop' into fb_amountsAndUnits

cnathe committed:
Measurement enum -> Unit / KindOfQuantity

labkey-susanh committed:
Convert units with alternate names to the standard base unit name

labkey-susanh commented on experiment/src/org/labkey/experiment/ExperimentUpgradeCode.java:
```diff
@@ -153,4 +186,166 @@ private static void ensureBigObjectIds(Selector lastValueSelector, Consumer<Long
+        else if (materialUnit != null && !materialUnit.isCompatible(baseUnit))
+        {
+            LOG.info("{} '{}' for sample '{}' is not compatible with the base unit '{}'. No conversion done.", isAliquot ? "Aliquot unit" : "Unit", materialUnit.name(), sampleMap.get(Name.name()), baseUnit);
+            sampleCounts.put("invalidUnits", sampleCounts.getOrDefault("invalidUnits", 0) + 1);
```
N.B. the capturing and logging of these counts will either get removed or have log level changed to DEBUG.

XingY replied:
Is it time to change those log level to DEBUG?

labkey-susanh replied:
Yes. Done.

cnathe committed:
Merge branch 'develop' into fb_amountsAndUnits

cnathe committed:
DataGenerator to use KindOfQuantity getCommonUnits()

cnathe committed:
SampleTypeAmountDisplayColumn fix for MSSQL rounding precision display

labkey-susanh committed:
Remove unnecessary map

labkey-susanh committed:
Give Count kind a name

labkey-susanh committed:
Register conversion types before the Tests run and small update to exception message

labkey-susanh committed:
Merge remote-tracking branch 'origin/develop' into fb_amountsAndUnits

labkey-susanh committed:
Move upgrade script to avoid conflict with script from develop

labkey-susanh committed:
Merge remote-tracking branch 'origin/develop' into fb_amountsAndUnits

cnathe committed:
StoredAmount column info to set default format for case without a sample type display unit

XingY commented on api/src/org/labkey/api/dataiterator/DetailedAuditLogDataIterator.java:
```diff
@@ -53,12 +55,14 @@ public enum AuditConfigs {
 
     // for batching
     final ArrayList<Map<String,Object>> _updatedRows = new ArrayList<>();
+    final ArrayList<Map<String, Object>> _providedValues = new ArrayList<>(); // TODO
```
Still TODO?

labkey-susanh replied:
Nope. Thanks.

labkey-susanh committed:
Add some logging, casting, and use a string instead of a Unit

XingY commented on experiment/src/org/labkey/experiment/ExperimentUpgradeCode.java:
```diff
@@ -153,4 +186,166 @@ private static void ensureBigObjectIds(Selector lastValueSelector, Consumer<Long
+        TableInfo tInfo = ExperimentService.get().getTinfoMaterial();
+        for (ExpSampleType sampleType :SampleTypeService.get().getSampleTypes(container, user, false))
+        {
+            try (DbScope.Transaction transaction = scope.ensureTransaction())
```
Why one transaction per container? If some container succeeded upgrade but some failed, won't that leave DB in a mixed state?

labkey-susanh replied:
Moved. This was largely due to an initial plan to have another pathway to use this code but that turned out to not be necessary.

XingY commented on experiment/src/org/labkey/experiment/ExperimentUpgradeCode.java:
```diff
@@ -153,4 +186,166 @@ private static void ensureBigObjectIds(Selector lastValueSelector, Consumer<Long
+                        .append(" WHERE cpastype = ?").add(sampleType.getLSID());
+                SqlSelector selector = new SqlSelector(scope, sql);
+                selector.mapStream().forEach(sampleMap -> {
+                    Map<String, Object> oldDataMap = new HashMap<>();
```
Could we use a prepared statement to do batch update?

labkey-susanh replied:
Yes. As discussed, I was wondering about the complications involved with different statements being needed depending on what was to be updated (I just opted for adding some no-op updates when fields are not changing) and the audit events that are necessary to be paired with each update statement, but there is already a method for prepared statements for audit events, so this route seems easier than originally thought.  Thanks.

XingY commented on api/src/org/labkey/api/query/AbstractQueryUpdateService.java:
```diff
@@ -719,6 +721,11 @@ protected Map<String, Object> coerceTypes(Map<String, Object> row)
                 {
                     if (PropertyType.FILE_LINK.equals(col.getPropertyType()))
                         value = ExpDataFileConverter.convert(value);
+                    else if (col.getKindOfQuantity() != null)
```
I thought with this story getKindOfQuantity() / getKindOfQuantity is now stale and show be removed? It's not clear to me how sample amount / unit is wired up to KindOfQuantity / Unit, seems it's just a one off hard wire?

labkey-susanh replied:
Sample Amount is not wired up to this at the moment, so really this code path won't get used yet.

labkey-susanh committed:
Still more logging

cnathe committed:
remove TODOs that are done

labkey-susanh committed:
Merge remote-tracking branch 'origin/fb_amountsAndUnits' into fb_amountsAndUnits

XingY commented on experiment/src/org/labkey/experiment/api/SampleTypeUpdateServiceDI.java:
```diff
@@ -557,6 +619,91 @@ public List<Map<String, Object>> updateRows(User user, Container container, List
+                    if (PropertyType.FILE_LINK.equals(col.getPropertyType()))
+                        value = ExpDataFileConverter.convert(value);
+                    else if (col.getKindOfQuantity() != null)
+                    {
```
I would think this can already handle amount/unit columns in the super class?

labkey-susanh replied:
This is just copying the logic from the overridden method.

XingY commented on experiment/src/org/labkey/experiment/api/SampleTypeUpdateServiceDI.java:
```diff
@@ -557,6 +619,91 @@ public List<Map<String, Object>> updateRows(User user, Container container, List
+        ColumnInfo unitsCol = null;
+        Object amountVal = null;
+        ColumnInfo amountCol = null;
+        if (row.containsKey(ExpMaterialTable.Column.Units.name()))
```
Not a fan of this overrride of deprecated method. Is there a way to have this in the super method?

labkey-susanh replied:
Not a fan either, but since this fix up is very specific to these two separate columns that are currently tied together, moving anything into the other method would be more messy.  Consider this transitional code that will go away either if we get rid of the need for the `coerceTypes` method used for one-at-a-time updates or we get to do more work so the `StoredAmount` field becomes a `Quantity`-type field and we then know how to deal with the amount and units together there.

XingY commented on experiment/src/org/labkey/experiment/api/SampleTypeUpdateServiceDI.java:
```diff
@@ -1922,43 +2077,109 @@ private void _addConvertColumn(ColumnInfo col, int fromIndex, int derivationData
 
         protected class SampleUnitsConvertColumn extends SimpleTranslator.SimpleConvertColumn
```
SampleUnitsConvertColumn and SampleAmountConvertColumn seems generic enough to be used for all volume/amount columns. Can they be moved to SimpleTranslator?

labkey-susanh replied:
Only sample types will have these columns. If we get to adding Quantity fields in other domains, we likely won't have a need for these convert columns (or they will be of a different nature).

XingY commented on experiment/src/org/labkey/experiment/ExperimentUpgradeCode.java:
```diff
@@ -153,4 +186,166 @@ private static void ensureBigObjectIds(Selector lastValueSelector, Consumer<Long
+    {
+        DbScope scope = ExperimentService.get().getSchema().getScope();
+        TableInfo tInfo = ExperimentService.get().getTinfoMaterial();
+        for (ExpSampleType sampleType :SampleTypeService.get().getSampleTypes(container, user, false))
```
Do we want to exclude specimen from this upgrade?

labkey-susanh replied:
I don't think it's necessary to exclude them. Since they don't have display units, very little will happen to the data.

XingY replied:
But it would make the upgrade faster. The query below is not cheap, and even though it's iterated as stream, it's still a lot of processing that could have been skipped.
`new SQLFragment("SELECT m.RowId, m.Name, m.StoredAmount, m.Units, m.AliquotVolume, m.AliquotUnit, m.AvailableAliquotVolume FROM ")
                        .append(tInfo, "m")
                        .append(" WHERE cpastype = ?").add(sampleType.getLSID());`

labkey-susanh replied:
Instead of doing any specialized treatment for specimen samples, I've updated the select statement to be more selective, excluding rows that will not need to be updated, which will probably eliminate further processing for specimen samples.

labkey-susanh committed:
Use LogHelper

XingY commented on experiment/src/org/labkey/experiment/ExperimentUpgradeCode.java:
```diff
@@ -153,4 +186,166 @@ private static void ensureBigObjectIds(Selector lastValueSelector, Consumer<Long
+                        newDataMap.put(Units.name(), null);
+                        sampleCounts.put("unitsWithoutAmounts", sampleCounts.getOrDefault("unitsWithoutAmounts", 0) + 1);
+                    }
+                    if (!StringUtils.isEmpty((String) sampleMap.get(AliquotUnit.name())) && sampleMap.get(AliquotVolume.name()) == null && sampleMap.get(AvailableAliquotVolume.name()) == null)
```
I think we just need to check for AliquotVolume. If AvailableAliquotVolume is not null, then AliquotVolume is for sure not null.

XingY replied:
But should we check for null, or >0 ? all samples (not aliquot) has not null AliquotVolume. It's initialized to 0.

labkey-susanh replied:
I would prefer to be more cautious with the checks here to be more sure of a consistent outcome. Even if the amount is 0, we may need to adjust the units value.

XingY reviewed (commented):
My question is, how is this worked tied to the previous work done to support ColumnRenderProperties.getKindOfQuantity

labkey-susanh committed:
Merge remote-tracking branch 'origin/develop' into fb_amountsAndUnits

labkey-susanh committed:
check for label as well as name in isExpMaterialColumn

labkey-susanh commented:
> My question is, how is this worked tied to the previous work done to support ColumnRenderProperties.getKindOfQuantity

As mentioned in standup, it's not really related to this other than using the `Quantity` class instead of `Measurement`. We are choosing to defer the single-column for a Quantity type field work until sometime later.

labkey-susanh committed:
Remove TODO

labkey-susanh committed:
Move transaction boundary

labkey-susanh committed:
Still more logging

labkey-susanh committed:
Remove assert that calls convert and finds the wrong units

labkey-susanh committed:
Move to debug logging

labkey-susanh committed:
Use prepared statement

labkey-susanh committed:
Remove some more logging

labkey-susanh committed:
Merge remote-tracking branch 'origin/develop' into fb_amountsAndUnits

labkey-susanh committed:
RawAmount and RawUnits should not be user-editable since they are dropped for data iterator

XingY commented on experiment/src/org/labkey/experiment/ExperimentUpgradeCode.java:
```diff
@@ -153,4 +194,215 @@ private static void ensureBigObjectIds(Selector lastValueSelector, Consumer<Long
+                    {
+                        updateStmt.addBatch();
+                        batchCount.getAndIncrement();
+                        SampleTimelineAuditEvent event = new SampleTimelineAuditEvent(container, "Storage amount unit conversion to base unit during upgrade script.");
```
Shouldn't the audit event be created at sample's container, rather than sample type's container?

labkey-susanh replied:
Yes. Thanks for that.

XingY commented on experiment/src/org/labkey/experiment/ExperimentUpgradeCode.java:
```diff
@@ -153,4 +194,215 @@ private static void ensureBigObjectIds(Selector lastValueSelector, Consumer<Long
+                            .append(" SET Units = ?, StoredAmount = ?, AliquotUnit = ?, AliquotVolume = ?, AvailableAliquotVolume = ? WHERE RowId = ?")
+                            .addAll(units, amount, aliquotUnits, aliquotAmount, availableAliquotAmount, rowId), null);
+
+            for (ExpSampleType sampleType : SampleTypeService.get().getSampleTypes(container, user, false))
```
Could we further optimize so we only process the sample type if either exist storedamount is not null OR aliquotvolume is not null?

labkey-susanh replied:
I don't think we can do anything at the sample type level, but I've updated the select statement for the samples in the sample type to exclude rows that won't need to be updated because they have neither amounts nor units.

XingY reviewed (commented):
upgrade script changes looks good, with just some further minor optimization comment.

labkey-susanh committed:
Use the sample's container for the upgrade audit event, if available, and skip the specimen sample type

labkey-susanh committed:
Rename scripts to make room for incoming

labkey-susanh committed:
Merge remote-tracking branch 'origin/develop' into fb_amountsAndUnits

labkey-susanh committed:
Update select statement to be more selective

labkey-susanh committed:
Don't add an audit event if the only thing that is updated is the AliquotUnits value

labkey-susanh committed:
Add comment

labkey-susanh committed:
Update schema version

XingY commented on experiment/src/org/labkey/experiment/ExperimentUpgradeCode.java:
```diff
@@ -153,4 +188,222 @@ private static void ensureBigObjectIds(Selector lastValueSelector, Consumer<Long
+                SQLFragment sql = new SQLFragment("SELECT m.RowId, m.Name, m.StoredAmount, m.Units, m.AliquotVolume, m.AliquotUnit, m.AvailableAliquotVolume, m.container FROM ")
+                        .append(tInfo, "m")
+                        .append(" WHERE cpastype = ?").add(sampleType.getLSID())
+                        .append(" AND (m.StoredAmount IS NOT NULL OR m.Units IS NOT NULL OR m.AliquotVolume IS NOT NULL OR m.AliquotUnit IS NOT NULL OR m.AvailableAliquotVolume IS NOT NULL)");
```
OR is generally slow. Could we at least get rid of AvailableAliquotVolume check? If AvailableAliquotVolume is not null, AliquotVolume is for sure not null, unless there is an App bug.

labkey-susanh replied:
We can certainly. Do you have any sense of how much difference the extra condition might make in terms of performance?  Yes, there are other problems if the sample row is in the state where AvailableAliquotVolume is the only thing that is not null in this set of fields, but if we don't do the update this might be a case of compounding an app bug. Again, I'm inclined to be conservative, particularly since this is upgrade code not run-all-the-time code.

XingY replied:
I remember in the past having to switch OR to UNION for performance reasons... And was advised to generally prefer UNION rather than OR. I don't know how to quantify the difference though.

labkey-susanh replied:
@labkey-matthewb Any comment here?

labkey-matthewb replied:
OR is especially bad when it is used across different tables e.g. t1.col IS NOT NULL OR t2.col IS NOT NULL.  In this case where we are probably just scanning the entire table anyway (and just once for upgrade), I think it probably is not a concern.  (expression evaluation is fast, cross-product join is very, very bad)

cnathe committed:
Set "RawAmount" column label to "Stored Amount" and "RawUnits" label to "Stored Units"

cnathe committed:
Merge branch 'develop' into fb_amountsAndUnits

XingY commented on experiment/src/org/labkey/experiment/ExperimentUpgradeCode.java:
```diff
@@ -153,4 +188,222 @@ private static void ensureBigObjectIds(Selector lastValueSelector, Consumer<Long
+                    if (batchCount.get() > 1000)
+                    {
+                        updateStmt.executeBatch();
+                        AuditLogService.get().addEvents(user, auditEvents, false);
```
So those events aren't really processed in a batch way, because AuditLogImpl._eventTypeQueue flattens the list to single a time. I think _eventTypeQueue should be updated to Queue<Pair<User, List<AuditTypeEvent>>>

labkey-susanh committed:
queue lists of events if we're given lists of events

labkey-susanh committed:
Don't use the "unit" Unit for AliquotUnit if there isn't a unit, just leave it as null

labkey-susanh committed:
Use the unit of the first aliquot when there is no other unit indicator

labkey-susanh committed:
Don't use base unit if there is no display unit.

labkey-susanh committed:
Revert change to use a List in the eventQueue since we aren't using the prepared statement anyway.

XingY commented on experiment/src/org/labkey/experiment/ExperimentUpgradeCode.java:
```diff
@@ -153,4 +188,202 @@ private static void ensureBigObjectIds(Selector lastValueSelector, Consumer<Long
+                Map<String, Integer> aliquotCounts = new HashMap<>();
+
+                Unit currentDisplayUnit = Unit.fromName(sampleType.getMetricUnit());
+                boolean hasDisplayUnit = currentDisplayUnit != null;
```
Looks like we can just skip the rest if hasDisplayUnit is false

cnathe committed:
Set default format for RawAmount column if none set, to match StoredAmount display column default

cnathe committed:
Merge branch 'develop' into fb_amountsAndUnits

cnathe committed:
Remove RawAmount and RawUnits labels
Add "Stored Amount" as an import alias for StoredAmount column
Add and use ImportAliasable.Helper.createImportMap() for SampleTypeUpdateServiceDI to get StoredAmount and Units column index

cnathe committed:
rename ImportAliasable createImportMap to createImportSet

cnathe committed:
remove Unit.count from getCommonUnits()

cnathe committed:
SampleTypeUpdateServiceDI update to fix warning for requesting RawAmount and RawUnits when selecting from exp.material on update from file

labkey-susanh committed:
Clear cache after doing the update

labkey-susanh committed:
Add space before units

labkey-susanh committed:
Update user display for upgrade audit event

labkey-susanh committed:
Merge branch 'fb_amountsAndUnits' of https://github.com/LabKey/platform into fb_amountsAndUnits

labkey-susanh committed:
Merge remote-tracking branch 'origin/develop' into fb_amountsAndUnits

labkey-susanh committed:
Add a warning message if we don't see the expected number of audit entries after upgrade

labkey-jeckels commented on experiment/src/org/labkey/experiment/ExperimentWarningProvider.java:
```diff
@@ -0,0 +1,58 @@
+public class ExperimentWarningProvider implements WarningProvider
+{
+    @Override
+    public void addDynamicWarnings(@NotNull Warnings warnings, @Nullable ViewContext context, boolean showAllWarnings)
```
I believe we'll call this method on every LKS page load for admins, so we should cache the message to avoid requerying (likely very large) audit table when there is a mismatch.

XingY replied:
One idea is to store an additional ExperimentModule.ACTUAL_AUDIT_COUNT_PROP in ExperimentModule.AMOUNT_AND_UNIT_UPGRADE_PROP property map.

labkey-susanh replied:
Storing the count as a class member seems to work.

labkey-jeckels commented on experiment/src/org/labkey/experiment/ExperimentWarningProvider.java:
```diff
@@ -0,0 +1,58 @@
+        if (props == null || props.isEmpty())
+            return;
+        String expectedCount = props.get("AuditRecordCount");
+        String transactionIdStr = props.get("AuditTransactionId");
```
Could use constants for these property names.

labkey-susanh replied:
Done

labkey-susanh committed:
Constants and caching

labkey-susanh committed:
Merge remote-tracking branch 'origin/develop' into fb_amountsAndUnits

labkey-susanh committed:
Add recommendation text

labkey-susanh committed:
Allow the property deletion to happen even from a GET page

XingY commented on api/src/org/labkey/api/ontology/Unit.java:
```diff
@@ -176,25 +185,56 @@ static Function<Double,Double> convertFn(Unit from, Unit to)
+        if (rawUnits == null)
+            return null;
+        if (rawUnits instanceof Unit u)
+            if (defaultUnits == null)
```
Can we add the outer braces to the block of if/else if/else ? It's confusing to read and error prone.

labkey-susanh replied:
Sure. Done.

XingY commented on experiment/src/org/labkey/experiment/ExperimentUpgradeCode.java:
```diff
@@ -153,4 +192,218 @@ private static void ensureBigObjectIds(Selector lastValueSelector, Consumer<Long
+            {
+                PropertyManager.WritablePropertyMap props = PropertyManager.getWritableProperties(AMOUNT_AND_UNIT_UPGRADE_PROP, true);
+                props.put(AUDIT_COUNT_PROP, auditCount.toString());
+                props.put(TRANSACTION_ID_PROP, transactionId == null ? null : transactionId.toString());
```
If transactionId is null, do we still want to set property map?

labkey-susanh replied:
It can't really be null unless my read of the logic is incorrect. This null check just to appease the IntelliJ warnings.

XingY commented on experiment/src/org/labkey/experiment/api/SampleTypeServiceImpl.java:
```diff
@@ -1542,37 +1538,31 @@ private Set<Long> getRootSampleIdsFromParentNames(String sampleTypeLsid, Set<Str
-        }
-
-        if (!StringUtils.isEmpty(totalDisplayUnitStr))
+        Unit totalUnit = null;
```
I feel we are not taking advantage of the fact now all aliquots are now stored in the same base "units", if sample type has a unit set. Rollup should now be much simpler (and faster) for sample types with unit, potentially could be done with a simple sql script.

labkey-susanh replied:
Yeah, I see what you're saying, but I'm going to not make this change now. We can make this optimization in a separate, smaller PR so we can concentrate on assuring no regressions with the rollup vs. the unit conversion code.

labkey-susanh replied:
Can you open an issue for this so we don't forget?

XingY replied:
https://www.labkey.org/home/Developer/issues/issues-details.view?issueId=53932

cnathe committed:
Merge branch 'develop' into fb_amountsAndUnits

labkey-susanh committed:
switch to debug logging

labkey-susanh committed:
braces for better readability
